<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#MovieLens-Case-Study" data-toc-modified-id="MovieLens-Case-Study-0.0.1">MovieLens Case Study</a></span></li><li><span><a href="#Datasets-Information:" data-toc-modified-id="Datasets-Information:-0.0.2">Datasets Information:</a></span></li><li><span><a href="#Objective:" data-toc-modified-id="Objective:-0.0.3">Objective:</a></span></li><li><span><a href="#Learning-Outcomes:" data-toc-modified-id="Learning-Outcomes:-0.0.4">Learning Outcomes:</a></span><ul class="toc-item"><li><span><a href="#Domain" data-toc-modified-id="Domain-0.0.4.1">Domain</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Data-Dictionary" data-toc-modified-id="Data-Dictionary-1">Data Dictionary</a></span><ul class="toc-item"><li><span><a href="#Ratings" data-toc-modified-id="Ratings-1.1">Ratings</a></span></li><li><span><a href="#Movie" data-toc-modified-id="Movie-1.2">Movie</a></span></li><li><span><a href="#User" data-toc-modified-id="User-1.3">User</a></span><ul class="toc-item"><li><span><a href="#1.-Import-the-necessary-packages" data-toc-modified-id="1.-Import-the-necessary-packages-1.3.1">1. Import the necessary packages</a></span></li><li><span><a href="#2.-Read-all-the-three-datasets" data-toc-modified-id="2.-Read-all-the-three-datasets-1.3.2">2. Read all the three datasets</a></span></li><li><span><a href="#3.-View-the-first-5-rows-of-all-the-datasets." data-toc-modified-id="3.-View-the-first-5-rows-of-all-the-datasets.-1.3.3">3. View the first 5 rows of all the datasets.</a></span></li><li><span><a href="#4.-Understand-the-shape-of-all-the-datasets." data-toc-modified-id="4.-Understand-the-shape-of-all-the-datasets.-1.3.4">4. Understand the shape of all the datasets.</a></span></li><li><span><a href="#5.-Check-the-data-types-of-the-columns-for-all-the-datasets." data-toc-modified-id="5.-Check-the-data-types-of-the-columns-for-all-the-datasets.-1.3.5">5. Check the data types of the columns for all the datasets.</a></span></li><li><span><a href="#6.-Give-a-statistical-summary-for-all-the-datasets." data-toc-modified-id="6.-Give-a-statistical-summary-for-all-the-datasets.-1.3.6">6. Give a statistical summary for all the datasets.</a></span></li><li><span><a href="#The-statistical-summary-of-the-data-will-give-us-an-overview-of-the-data-by-providing-key-statistics-of-the-different-columns-in-the-data." data-toc-modified-id="The-statistical-summary-of-the-data-will-give-us-an-overview-of-the-data-by-providing-key-statistics-of-the-different-columns-in-the-data.-1.3.7">The statistical summary of the data will give us an overview of the data by providing key statistics of the different columns in the data.</a></span></li><li><span><a href="#7.-Find-the-number-of-movies-per-genre-using-the-movie-data" data-toc-modified-id="7.-Find-the-number-of-movies-per-genre-using-the-movie-data-1.3.8">7. Find the number of movies per genre using the movie data</a></span></li><li><span><a href="#Applying-functions-on-specific-columns-of-the-data-will-help-us-in-gaining-specific-insights-about-the-data." data-toc-modified-id="Applying-functions-on-specific-columns-of-the-data-will-help-us-in-gaining-specific-insights-about-the-data.-1.3.9">Applying functions on specific columns of the data will help us in gaining specific insights about the data.</a></span></li><li><span><a href="#8.-Find-the-movies-that-have-more-than-one-genre" data-toc-modified-id="8.-Find-the-movies-that-have-more-than-one-genre-1.3.10">8. Find the movies that have more than one genre</a></span></li><li><span><a href="#Creating-new-columns-using-existing-columns-in-the-dataframe-requires-domain-knowledge,-and-can-help-us-to-understand-our-data-better." data-toc-modified-id="Creating-new-columns-using-existing-columns-in-the-dataframe-requires-domain-knowledge,-and-can-help-us-to-understand-our-data-better.-1.3.11">Creating new columns using existing columns in the dataframe requires domain knowledge, and can help us to understand our data better.</a></span></li><li><span><a href="#9.-Find-the-top-25-movies-according-to-average-ratings-such-that-each-movie-has-number-of-ratings-more-than-100" data-toc-modified-id="9.-Find-the-top-25-movies-according-to-average-ratings-such-that-each-movie-has-number-of-ratings-more-than-100-1.3.12">9. Find the top 25 movies according to average ratings such that each movie has number of ratings more than 100</a></span></li><li><span><a href="#We-have-just-learned-how-to-use-groupby()-to--group-the-various-categories-in-the-data,-and-how-to-use-sort_values()-to-sort-the-data." data-toc-modified-id="We-have-just-learned-how-to-use-groupby()-to--group-the-various-categories-in-the-data,-and-how-to-use-sort_values()-to-sort-the-data.-1.3.13">We have just learned how to use groupby() to  group the various categories in the data, and how to use sort_values() to sort the data.</a></span></li><li><span><a href="#10.-Find-the-relationship-between-user-demographics-and-movie-ratings" data-toc-modified-id="10.-Find-the-relationship-between-user-demographics-and-movie-ratings-1.3.14">10. Find the relationship between user demographics and movie ratings</a></span><ul class="toc-item"><li><span><a href="#10.1-Find-the-mean-rating-for-every-occupation" data-toc-modified-id="10.1-Find-the-mean-rating-for-every-occupation-1.3.14.1">10.1 Find the mean rating for every occupation</a></span></li><li><span><a href="#10.2-Find-mean-rating-for-each-gender" data-toc-modified-id="10.2-Find-mean-rating-for-each-gender-1.3.14.2">10.2 Find mean rating for each gender</a></span></li><li><span><a href="#10.3-Find-the-mean-rating-grouped-by-both-occupation-and-gender" data-toc-modified-id="10.3-Find-the-mean-rating-grouped-by-both-occupation-and-gender-1.3.14.3">10.3 Find the mean rating grouped by both occupation and gender</a></span></li></ul></li><li><span><a href="#Using-groupby()-on-multiple-columns-of-the-data-helps-us-analyze-subsets-of-the-data-to-gain-insights." data-toc-modified-id="Using-groupby()-on-multiple-columns-of-the-data-helps-us-analyze-subsets-of-the-data-to-gain-insights.-1.3.15">Using groupby() on multiple columns of the data helps us analyze subsets of the data to gain insights.</a></span></li><li><span><a href="#11.-Find-the-mean-rating-for-each-genre." data-toc-modified-id="11.-Find-the-mean-rating-for-each-genre.-1.3.16">11. Find the mean rating for each genre.</a></span></li><li><span><a href="#12.-Divide-the-age-column-into-four-groups." data-toc-modified-id="12.-Divide-the-age-column-into-four-groups.-1.3.17">12. Divide the age column into four groups.</a></span><ul class="toc-item"><li><span><a href="#12.1-Find-the-mean-rating-for-each-age-group." data-toc-modified-id="12.1-Find-the-mean-rating-for-each-age-group.-1.3.17.1">12.1 Find the mean rating for each age group.</a></span></li><li><span><a href="#12.2-Find-the-mean,-median,-and-standard-deviation-of-rating-for-each-age-group." data-toc-modified-id="12.2-Find-the-mean,-median,-and-standard-deviation-of-rating-for-each-age-group.-1.3.17.2">12.2 Find the mean, median, and standard deviation of rating for each age group.</a></span></li><li><span><a href="#12.3-Which-genre-movies-are-common-among-Engineers?" data-toc-modified-id="12.3-Which-genre-movies-are-common-among-Engineers?-1.3.17.3">12.3 Which genre movies are common among Engineers?</a></span></li></ul></li><li><span><a href="#We-have-just-learned-how-to-use-the-pandas-cut()-function-to-group-a-numerical-column-in-the-data-into-different-categories-and-use-agg()-to-use-multiple-aggregate-functions-at-once-with-groupby().-These-are-important-techniques-that-allow-us-to-create-informative-groups-from-numerical-columns-and-understand-multiple-statistics-of-subsets-of-our-data-at-once." data-toc-modified-id="We-have-just-learned-how-to-use-the-pandas-cut()-function-to-group-a-numerical-column-in-the-data-into-different-categories-and-use-agg()-to-use-multiple-aggregate-functions-at-once-with-groupby().-These-are-important-techniques-that-allow-us-to-create-informative-groups-from-numerical-columns-and-understand-multiple-statistics-of-subsets-of-our-data-at-once.-1.3.18">We have just learned how to use the pandas cut() function to group a numerical column in the data into different categories and use agg() to use multiple aggregate functions at once with groupby(). These are important techniques that allow us to create informative groups from numerical columns and understand multiple statistics of subsets of our data at once.</a></span></li></ul></li></ul></li></ul></div>

### MovieLens Case Study

The GroupLens Research Project is a research group in the Department of Computer Science and Engineering at the University of Minnesota. The data is widely used for collaborative filtering and other filtering solutions. However, we will be using this data to act as a means to demonstrate our skill in using Python to “play” with data.

### Datasets Information:

- **ratings.csv:** It contains information on ratings given by the users to a particular movie. Columns: user id, movie id, rating, timestamp

- **movie.csv:** The file contains information related to the movies and their genre. Columns: movie id, movie title, release date, unknown, Action, Adventure, Animation, Children’s, Comedy, Crime, Documentary, Drama, Fantasy, Film-Noir, Horror, Musical, Mystery, Romance, Sci-Fi, Thriller, War, Western

- **user.csv:** It contains information about the users who have rated the movies.. Columns: user id, age, gender, occupation, zip code

### Objective:

`To extract insights from the dataset`

### Learning Outcomes:
`Use of Pandas Functions - shape, describe, groupby, merge etc.`


#### Domain 
`Internet and Entertainment`

**Note that the case study will need you to apply the concepts of groupby and merging extensively.**

# Data Dictionary

## Ratings
-

## Movie

## User

### 1. Import the necessary packages

In [1]:
import pandas as pd
import numpy as np

### 2. Read all the three datasets

In [2]:
# Reading datasets by using read_csv from pandas package
ratings = pd.read_csv("../data/ratings.csv")
movie = pd.read_csv("../data/movie.csv")
user = pd.read_csv("../data/user.csv")

### 3. View the first 5 rows of all the datasets.
`Note that you will need to do it for all three datasets separately`

In [3]:
ratings.head(5)

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [4]:
movie.head(5)

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [5]:
user.head(5)

user id  age gender  occupation zip code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

### 4. Understand the shape of all the datasets.
`Note that you will need to do it for all the three datasets seperately`

In [6]:
# ratings
ratings.shape

(100000, 4)

 **Observation:** There are 100000 rows and 4 columns in the ratings dataset

In [7]:
# user
user.shape

(943, 5)

 **Observation:** There are 943 rows and 5 columns in the user dataset

In [8]:
# movie
movie.shape

(1680, 21)

 **Observation:** There are 1680 rows and 21 columns in the movie dataset

### 5. Check the data types of the columns for all the datasets.
 `Note that you will need to do it for all three datasets separately`

In [9]:
# ratings
# We use dataframe.dtypes to get the data types of each column
ratings.dtypes 

user id      int64
movie id     int64
rating       int64
timestamp    int64
dtype: object

 **Observation:** All columns have integer data type 

In [10]:
# user
user.dtypes

user id        int64
age            int64
gender        object
occupation    object
zip code      object
dtype: object

 **Observations:**
 1. user id and age columns are of integer data types 
 2. gender, occupation and zip code columns are of string data type

In [11]:
# movie
movie.dtypes

movie id         int64
movie title     object
release date    object
Action           int64
Adventure        int64
Animation        int64
Childrens        int64
Comedy           int64
Crime            int64
Documentary      int64
Drama            int64
Fantasy          int64
Film-Noir        int64
Horror           int64
Musical          int64
Mystery          int64
Romance          int64
Sci-Fi           int64
Thriller         int64
War              int64
Western          int64
dtype: object

**Observation:**
1. movie title and release date are of string data type
2. movie id and all genres are of interger data type

### 6. Give a statistical summary for all the datasets.
`Note that you will need to do it for all three datasets separately`

In [12]:
# ratings
ratings.describe()

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

**Observation:** The mean and Median of user ratings are 3.53 & 4.00 respectively

In [13]:
# user
user.describe()

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

**Observation:** The average age of all the users is 34 years while the range lies between 7 to 73 years.

In [14]:
# movie
movie.describe()

movie id       Action    Adventure    Animation    Childrens  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean    841.525595     0.149405     0.080357     0.025000     0.072619   
std     485.609591     0.356593     0.271926     0.156171     0.259587   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%     421.750000     0.000000     0.000000     0.000000     0.000000   
50%     841.500000     0.000000     0.000000     0.000000     0.000000   
75%    1261.250000     0.000000     0.000000     0.000000     0.000000   
max    1682.000000     1.000000     1.000000     1.000000     1.000000   

            Comedy        Crime  Documentary        Drama      Fantasy  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.300595     0.064881     0.029762     0.431548     0.013095   
std       0.458653     0.246389     0.169980     0.495440     0.113717   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         Film-Noir       Horror      Musical      Mystery      Romance  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.014286     0.054762     0.033333     0.036310     0.147024   
std       0.118701     0.227583     0.179559     0.187115     0.354235   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

            Sci-Fi     Thriller          War      Western  
count  1680.000000  1680.000000  1680.000000  1680.000000  
mean      0.060119     0.149405     0.042262     0.016071  
std       0.237778     0.356593     0.201246     0.125788  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000

In [1]:
movie.isna().sum()

NameError: name 'movie' is not defined

**Observation:** The genres should be in categorical format and not in the numeric because it is of binary class

### The statistical summary of the data will give us an overview of the data by providing key statistics of the different columns in the data.

### 7. Find the number of movies per genre using the movie data

In [16]:
# Getting all the column names
movie.columns

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [17]:
# Taking all the genre columns and finding the sum for every column
movie[[ 'Action',
       'Adventure', 'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary',
       'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery',
       'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']].sum()

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [18]:
# Alternatively, we can also loc function
movie.loc[:,'Action':'Western'].sum()

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [19]:
# Sorting the movies across genres
number = movie.loc[:,'Action':'Western'].sum()
number.sort_values(ascending = False)

Drama          725
Comedy         505
Action         251
Thriller       251
Romance        247
Adventure      135
Childrens      122
Crime          109
Sci-Fi         101
Horror          92
War             71
Mystery         61
Musical         56
Documentary     50
Animation       42
Western         27
Film-Noir       24
Fantasy         22
dtype: int64

**Observations:**
1. Drama and Comedy are the most common movie genre.
2. Clearly, there are some movies that have more than one genre.

### Applying functions on specific columns of the data will help us in gaining specific insights about the data.

### 8. Find the movies that have more than one genre
`Hint: use sum on the axis = 1`

In [20]:
# Checking column names
movie.columns

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [21]:
# we create a new dataframe using two columns of the movie dataframe
new_movie = movie[['movie id', 'movie title']].copy()

In [22]:
new_movie["Number of Genres"] = movie.loc[:, 'Action':'Western'].sum(axis=1)

In [23]:
# Filtering movies that have more than 1 genres
new_movie[new_movie['Number of Genres']>1]

movie id         movie title  Number of Genres
0            1          Toy Story                  3
1            2          GoldenEye                  3
3            4         Get Shorty                  3
4            5            Copycat                  3
6            7     Twelve Monkeys                  2
...        ...                 ...               ...
1666      1669  MURDER and murder                  3
1667      1670            Tainted                  2
1670      1673             Mirage                  2
1676      1679          B. Monkey                  2
1677      1680      Sliding Doors                  2

[849 rows x 3 columns]

In [24]:
new_movie[new_movie['Number of Genres']>1].reset_index(drop = True)

movie id         movie title  Number of Genres
0           1          Toy Story                  3
1           2          GoldenEye                  3
2           4         Get Shorty                  3
3           5            Copycat                  3
4           7     Twelve Monkeys                  2
..        ...                 ...               ...
844      1669  MURDER and murder                  3
845      1670            Tainted                  2
846      1673             Mirage                  2
847      1679          B. Monkey                  2
848      1680      Sliding Doors                  2

[849 rows x 3 columns]

**Observation:** 849 movies have more than one genre.

### Creating new columns using existing columns in the dataframe requires domain knowledge, and can help us to understand our data better.

### 9. Find the top 25 movies according to average ratings such that each movie has number of ratings more than 100

Hint : 

1. First find the movies that have more than 100 ratings(use groupby and count). Extract the movie id in a list.
2. Find the average rating of all the movies and sort them in the descending order. 
3. Use isin(list obtained from 1) to filter out the movies which have more than 100 ratings.
4. You will have to use the .merge() function to get the movie titles.

Note: This question will need you to research about groupby and apply your findings. You can find more on groupby on https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html.

In [25]:
# Merging ratings dataset with movie dataset
df_merge = movie.merge(ratings, on = 'movie id', how = 'inner')
df_merge.head()

movie id movie title release date  Action  Adventure  Animation  Childrens  \
0         1  Toy Story      1-Jan-95       0          0          1          1   
1         1  Toy Story      1-Jan-95       0          0          1          1   
2         1  Toy Story      1-Jan-95       0          0          1          1   
3         1  Toy Story      1-Jan-95       0          0          1          1   
4         1  Toy Story      1-Jan-95       0          0          1          1   

   Comedy  Crime  Documentary  ...  Musical  Mystery  Romance  Sci-Fi  \
0       1      0            0  ...        0        0        0       0   
1       1      0            0  ...        0        0        0       0   
2       1      0            0  ...        0        0        0       0   
3       1      0            0  ...        0        0        0       0   
4       1      0            0  ...        0        0        0       0   

   Thriller  War  Western  user id  rating  timestamp  
0         0    0        0      308       4  887736532  
1         0    0        0      287       5  875334088  
2         0    0        0      148       4  877019411  
3         0    0        0      280       4  891700426  
4         0    0        0       66       3  883601324  

[5 rows x 24 columns]

In [26]:
# Checking the dimensions of the merged dataframe
df_merge.shape

(99990, 24)

In [27]:
# Finding the count of ratings for each movie using groupby() and count()
# reset_index() is used to shift movie title from being the dataframe’s (movie_count’s) index to 
# being just a normal column 
movie_count = df_merge.groupby(['movie title'])['rating'].count().reset_index()
movie_count.head()

movie title  rating
0  'Til There Was You        9
1               1-900        5
2      101 Dalmatians      109
3        12 Angry Men      125
4                  187      41

In [28]:
# Extracting the movie titles that have more than 100 ratings 
movie_100 = movie_count[movie_count['rating']>100]['movie title']
movie_100.head()

2            101 Dalmatians 
3              12 Angry Men 
7     2001: A Space Odyssey 
15           Absolute Power 
16               Abyss, The 
Name: movie title, dtype: object

In [49]:
df_merge.head()

movie id movie title release date  Action  Adventure  Animation  Childrens  \
0         1  Toy Story      1-Jan-95       0          0          1          1   
1         1  Toy Story      1-Jan-95       0          0          1          1   
2         1  Toy Story      1-Jan-95       0          0          1          1   
3         1  Toy Story      1-Jan-95       0          0          1          1   
4         1  Toy Story      1-Jan-95       0          0          1          1   

   Comedy  Crime  Documentary  ...  Musical  Mystery  Romance  Sci-Fi  \
0       1      0            0  ...        0        0        0       0   
1       1      0            0  ...        0        0        0       0   
2       1      0            0  ...        0        0        0       0   
3       1      0            0  ...        0        0        0       0   
4       1      0            0  ...        0        0        0       0   

   Thriller  War  Western  user id  rating  timestamp  
0         0    0        0      308       4  887736532  
1         0    0        0      287       5  875334088  
2         0    0        0      148       4  877019411  
3         0    0        0      280       4  891700426  
4         0    0        0       66       3  883601324  

[5 rows x 24 columns]

In [56]:
df_merge.groupby(['movie title'], as_index = False)['rating'].mean().sort_values(ascending=False, by = "rating")

movie title  rating
631                       Great Day in Harlem, A      5.0
1170                                 Prefontaine      5.0
1354                      Someone Else's America      5.0
461   Entertaining Angels: The Dorothy Day Story      5.0
942          Marlene Dietrich: Shadow and Light (     5.0
...                                           ...     ...
111                                    Babyfever      1.0
834                               Lashou shentan      1.0
1307                            Shadows (Cienie)      1.0
1304       Shadow of Angels (Schatten der Engel)      1.0
1168                                    Power 98      1.0

[1657 rows x 2 columns]

In [29]:
# Finding average ratings for each movie and sorting them out in descending order
# using groupby() and sort_values() on merged data frame
avg_rating = df_merge.groupby(['movie title'])['rating'].mean().sort_values(ascending=False).reset_index()
avg_rating

movie title  rating
0                  Saint of Fort Washington, The      5.0
1     Entertaining Angels: The Dorothy Day Story      5.0
2                                  Aiqing wansui      5.0
3                             Santa with Muscles      5.0
4                                    Prefontaine      5.0
...                                           ...     ...
1652         Tigrero: A Film That Was Never Made      1.0
1653                               Low Life, The      1.0
1654                              Pharaoh's Army      1.0
1655                        To Cross the Rubicon      1.0
1656                                    Power 98      1.0

[1657 rows x 2 columns]

In [62]:
# Extracting movie titles that have more than 100 ratings using movie titles in movie_100 and isin() function
# Displaying top 25 rows only
final_dataset = avg_rating[avg_rating['movie title'].isin(movie_100)]
final_dataset

movie title    rating
15               Close Shave, A   4.491071
16             Schindler's List   4.466443
17          Wrong Trousers, The   4.466102
18                   Casablanca   4.456790
20    Shawshank Redemption, The   4.445230
...                          ...       ...
1261                      Spawn   2.615385
1277              Event Horizon   2.574803
1291                      Crash   2.546875
1357              Jungle2Jungle   2.439394
1392             Cable Guy, The   2.339623

[334 rows x 2 columns]

In [97]:
final_dataset[final_dataset['movie title']== 'Star Wars ']

movie title    rating
23  Star Wars   4.358491

### We have just learned how to use groupby() to  group the various categories in the data, and how to use sort_values() to sort the data.

### 10. Find the relationship between user demographics and movie ratings

#### 10.1 Find the mean rating for every occupation 

1. Merge the user dataset with movie and ratings(already merged : df_merge) dataset
2. Apply groupby on occupation

In [31]:
# Merging user dataset with movie and ratings(already merged : df_merge) dataset
df_merge_all = df_merge.merge(user, on = 'user id', how = 'inner')

In [65]:
# Group by occupation and aggregate with mean
df_merge_all.groupby('occupation', as_index = False).rating.mean()

occupation    rating
0   administrator  3.635646
1          artist  3.653380
2          doctor  3.688889
3        educator  3.670692
4        engineer  3.541473
5   entertainment  3.440783
6       executive  3.349794
7      healthcare  2.896220
8       homemaker  3.301003
9          lawyer  3.735316
10      librarian  3.560781
11      marketing  3.485641
12           none  3.777778
13          other  3.552335
14     programmer  3.568205
15        retired  3.466750
16       salesman  3.582944
17      scientist  3.611273
18        student  3.515167
19     technician  3.532097
20         writer  3.376152

**Observation:**: The mean rating is low (2.89) for people working in healthcare.

#### 10.2 Find mean rating for each gender

In [33]:
df_merge_all.groupby('gender').rating.mean()

gender
F    3.531510
M    3.529333
Name: rating, dtype: float64

**Observation**: The ratings given by men and women are similar on average.

#### 10.3 Find the mean rating grouped by both occupation and gender
1. Put both the columns under groupby function and find the mean rating. 

In [34]:
df_merge_all.groupby(['occupation','gender']).rating.mean()

occupation     gender
administrator  F         3.781839
               M         3.555233
artist         F         3.347065
               M         3.875841
doctor         M         3.688889
educator       F         3.699132
               M         3.660246
engineer       F         3.751724
               M         3.537676
entertainment  F         3.448889
               M         3.439807
executive      F         3.773756
               M         3.320340
healthcare     F         2.736021
               M         3.639839
homemaker      F         3.278810
               M         3.500000
lawyer         F         3.623188
               M         3.741379
librarian      F         3.580070
               M         3.537920
marketing      F         3.522624
               M         3.474801
none           F         3.632877
               M         3.876636
other          F         3.531114
               M         3.563447
programmer     F         3.577566
               M         3

**Observations:** 

1. We had seen that the healthcare sector had a low mean rating. We now see that the low rating is driven by the female workers in the healthcare sector.

2. The male writers have given lower ratings compared to the female writers.

3. Only male doctors have given ratings. No female doctors have given ratings as per the data.

### Using groupby() on multiple columns of the data helps us analyze subsets of the data to gain insights.

### 11. Find the mean rating for each genre.

In [66]:
cols

Index(['Action', 'Adventure', 'Animation', 'Childrens', 'Comedy', 'Crime',
       'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical',
       'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [67]:
df_merge_all

movie id           movie title release date  Action  Adventure  \
0             1            Toy Story      1-Jan-95       0          0   
1             4           Get Shorty      1-Jan-95       1          0   
2             5              Copycat      1-Jan-95       0          0   
3             7       Twelve Monkeys      1-Jan-95       0          0   
4             8                 Babe      1-Jan-95       0          0   
...         ...                   ...          ...     ...        ...   
99985       748           Saint, The     14-Mar-97       1          0   
99986       751  Tomorrow Never Dies      1-Jan-97       1          0   
99987       879      Peacemaker, The      1-Jan-97       1          0   
99988       894         Home Alone 3      1-Jan-97       0          0   
99989       901            Mr. Magoo     25-Dec-97       0          0   

       Animation  Childrens  Comedy  Crime  Documentary  ...  War  Western  \
0              1          1       1      0            0  ...    0        0   
1              0          0       1      0            0  ...    0        0   
2              0          0       0      1            0  ...    0        0   
3              0          0       0      0            0  ...    0        0   
4              0          1       1      0            0  ...    0        0   
...          ...        ...     ...    ...          ...  ...  ...      ...   
99985          0          0       0      0            0  ...    0        0   
99986          0          0       0      0            0  ...    0        0   
99987          0          0       0      0            0  ...    1        0   
99988          0          1       1      0            0  ...    0        0   
99989          0          0       1      0            0  ...    0        0   

       user id  rating  timestamp  age  gender  occupation  zip code  \
0          308       4  887736532   60       M     retired     95076   
1          308       5  887737890   60       M     retired     95076   
2          308       4  887739608   60       M     retired     95076   
3          308       4  887738847   60       M     retired     95076   
4          308       5  887736696   60       M     retired     95076   
...        ...     ...        ...  ...     ...         ...       ...   
99985      729       4  893286638   19       M     student     56567   
99986      729       3  893286338   19       M     student     56567   
99987      729       3  893286299   19       M     student     56567   
99988      729       1  893286511   19       M     student     56567   
99989      729       1  893286491   19       M     student     56567   

       age_group  
0        Elderly  
1        Elderly  
2        Elderly  
3        Elderly  
4        Elderly  
...          ...  
99985   Teenager  
99986   Teenager  
99987   Teenager  
99988   Teenager  
99989   Teenager  

[99990 rows x 29 columns]

In [69]:
cols=df_merge_all.loc[:,'Action':'Western'].columns

for i in cols:
    print(i,':' , np.round(df_merge_all[df_merge_all[i]==1].rating.mean(),2))

Action : 3.48
Adventure : 3.5
Animation : 3.58
Childrens : 3.35
Comedy : 3.39
Crime : 3.63
Documentary : 3.67
Drama : 3.69
Fantasy : 3.22
Film-Noir : 3.92
Horror : 3.29
Musical : 3.52
Mystery : 3.64
Romance : 3.62
Sci-Fi : 3.56
Thriller : 3.51
War : 3.82
Western : 3.61


**Observation**: War movies have the highest average rating, while Fantasy movies have the lowest average rating.

### 12. Divide the age column into four groups.

* Group 1. 0-20 => Teenager
* Group 2. 20-40 => Adult
* Group 3. 40-55 => Middle Age
* Group 4. 55+ => Elderly

This questions requires pandas cut() function.

Refer this link for more info: https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.cut.html

In [90]:
df_merge['movie title'].nunique()

1657

In [95]:
(df_merge[(df_merge['movie title'] == 'Star Wars ')])

movie id movie title release date  Action  Adventure  Animation  \
6402        50  Star Wars      1-Jan-77       1          1          0   
6403        50  Star Wars      1-Jan-77       1          1          0   
6404        50  Star Wars      1-Jan-77       1          1          0   
6405        50  Star Wars      1-Jan-77       1          1          0   
6406        50  Star Wars      1-Jan-77       1          1          0   
...        ...         ...          ...     ...        ...        ...   
6980        50  Star Wars      1-Jan-77       1          1          0   
6981        50  Star Wars      1-Jan-77       1          1          0   
6982        50  Star Wars      1-Jan-77       1          1          0   
6983        50  Star Wars      1-Jan-77       1          1          0   
6984        50  Star Wars      1-Jan-77       1          1          0   

      Childrens  Comedy  Crime  Documentary  ...  Musical  Mystery  Romance  \
6402          0       0      0            0  ...        0        0        1   
6403          0       0      0            0  ...        0        0        1   
6404          0       0      0            0  ...        0        0        1   
6405          0       0      0            0  ...        0        0        1   
6406          0       0      0            0  ...        0        0        1   
...         ...     ...    ...          ...  ...      ...      ...      ...   
6980          0       0      0            0  ...        0        0        1   
6981          0       0      0            0  ...        0        0        1   
6982          0       0      0            0  ...        0        0        1   
6983          0       0      0            0  ...        0        0        1   
6984          0       0      0            0  ...        0        0        1   

      Sci-Fi  Thriller  War  Western  user id  rating  timestamp  
6402       1         0    1        0      290       5  880473582  
6403       1         0    1        0       79       4  891271545  
6404       1         0    1        0        2       5  888552084  
6405       1         0    1        0        8       5  879362124  
6406       1         0    1        0      274       5  878944679  
...      ...       ...  ...      ...      ...     ...        ...  
6980       1         0    1        0      749       5  878846978  
6981       1         0    1        0      886       5  876031501  
6982       1         0    1        0      739       4  886958895  
6983       1         0    1        0      903       5  891031329  
6984       1         0    1        0      780       5  891363685  

[583 rows x 24 columns]

In [96]:
df_merge[['movie title', 'rating']]['movie title'].value_counts().reset_index()

index  movie title
0                                Star Wars           583
1                                  Contact           509
2                                    Fargo           508
3                       Return of the Jedi           507
4                                Liar Liar           485
...                                     ...          ...
1652                          Two Friends (            1
1653                              Dadetown             1
1654                              Liebelei             1
1655  Land and Freedom (Tierra y libertad)             1
1656                      Modern Affair, A             1

[1657 rows x 2 columns]

In [ ]:
movie_count1 = df_merge[['movie title', 'rating']].value_counts()
movie_count1.head()

In [36]:
#create a new column with name age_group
df_merge_all['age_group']=pd.cut(df_merge_all.age,bins=(0,20,40,55,100),labels=('Teenager','Adult','Middle Age','Elderly'))

In [82]:
df_merge_all.age.unique()

array([60, 21, 33, 30, 23, 29, 26, 31, 24, 32, 50, 17, 45, 20, 11, 19, 25,
       46, 38, 59, 61, 42, 28, 57, 27, 40, 35, 52, 51, 49, 22, 36, 53, 43,
       15, 48, 18, 44, 10, 39, 47, 41, 34, 37, 54, 63, 13, 14, 16, 58, 70,
        7, 55, 69, 73, 65, 56, 68, 66, 64, 62])

In [37]:
#print top 5 entries of the age and age_group columns
df_merge_all[['age','age_group']].head()

age age_group
0   60   Elderly
1   60   Elderly
2   60   Elderly
3   60   Elderly
4   60   Elderly

In [38]:
#print last 5 entries of the age and age_group columns
df_merge_all[['age','age_group']].tail()

age age_group
99985   19  Teenager
99986   19  Teenager
99987   19  Teenager
99988   19  Teenager
99989   19  Teenager

#### 12.1 Find the mean rating for each age group.

In [70]:
np.round(df_merge_all.groupby(['age_group']).rating.mean(),2)

age_group
Teenager      3.55
Adult         3.49
Middle Age    3.61
Elderly       3.65
Name: rating, dtype: float64

**Observation:** Old people have given higher ratings than any other age group.

#### 12.2 Find the mean, median, and standard deviation of rating for each age group.

1. This can be done using groupby.agg(), where agg() stands for aggregate.
2. agg() can take more than 1 aggregate function simultaneously.

Refer to this link for more info :https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.DataFrameGroupBy.aggregate.html

In [72]:
df_merge_all.groupby(['age_group']).rating.agg(['mean','median','std']).reset_index()

age_group      mean  median       std
0    Teenager  3.548290       4  1.154198
1       Adult  3.492513       4  1.138085
2  Middle Age  3.606980       4  1.083602
3     Elderly  3.651703       4  1.030762

#### 12.3 Which genre movies are common among Engineers?

In [73]:
df_merge_all.groupby('occupation').sum().loc[:,'Action':'Western'].loc['engineer']

Action         2277
Adventure      1299
Animation       363
Childrens       629
Comedy         2438
Crime           611
Documentary      51
Drama          3153
Fantasy         125
Film-Noir       145
Horror          414
Musical         441
Mystery         363
Romance        1467
Sci-Fi         1221
Thriller       1712
War             858
Western         180
Name: engineer, dtype: int64

**Observation:**   Engineers mostly watch movies of Drama, Comedy, and Action genres.

You can try finding the most common genre among different professionals.

### We have just learned how to use the pandas cut() function to group a numerical column in the data into different categories and use agg() to use multiple aggregate functions at once with groupby(). These are important techniques that allow us to create informative groups from numerical columns and understand multiple statistics of subsets of our data at once.